esp32s2-hal/examples/spi_slave_dma.rs

```diff
@@ -0,0 +1,171 @@
+        &clocks,
+        &mut system.peripheral_clock_control,
+    );
+    let mut wdt0 = timer_group0.wdt;
+
+    rtc.rwdt.disable();
+    wdt0.disable();
+
+    let io = IO::new(peripherals.GPIO, peripherals.IO_MUX);
+    let slave_sclk = io.pins.gpio6;
+    let mut master_sclk = io.pins.gpio5.into_push_pull_output();
+    let slave_miso = io.pins.gpio2;
+    let master_miso = io.pins.gpio1.into_floating_input();
+    let slave_mosi = io.pins.gpio7;
+    let mut master_mosi = io.pins.gpio8.into_push_pull_output();
+    let slave_cs = io.pins.gpio10;
+    let mut master_cs = io.pins.gpio9.into_push_pull_output();
+    master_cs.set_high().unwrap();
+    master_sclk.set_low().unwrap();
+    master_mosi.set_low().unwrap();
+
+    let dma = Dma::new(system.dma, &mut system.peripheral_clock_control);
+    let dma_channel = dma.spi2channel;
+
+    let mut descriptors = [0u32; 8 * 3];
+    let mut rx_descriptors = [0u32; 8 * 3];
+
+    let mut spi = Spi::new(
+        peripherals.SPI2,
+        slave_sclk,
+        slave_mosi,
+        slave_miso,
+        slave_cs,
+        SpiMode::Mode0,
+        &mut system.peripheral_clock_control,
+    )
+    .with_dma(dma_channel.configure(
+        false,
+        &mut descriptors,
+        &mut rx_descriptors,
+        DmaPriority::Priority0,
+    ));
+
+    let mut delay = Delay::new(&clocks);
+
+    // DMA buffer require a static life-time
+    let master_send = &mut [0u8; 32000];
+    let master_receive = &mut [0u8; 32000];
+    let mut slave_send = buffer1();
+    let mut slave_receive = buffer2();
+    let mut i = 0;
+
+    for (i, v) in master_send.iter_mut().enumerate() {
+        *v = (i % 255) as u8;
+    }
+    for (i, v) in slave_send.iter_mut().enumerate() {
+        *v = (254 - (i % 255)) as u8;
+    }
+
+    loop {
+        master_send[0] = i;
+        master_send[master_send.len() - 1] = i;
+        slave_send[0] = i;
+        slave_send[slave_send.len() - 1] = i;
+        slave_receive.fill(0xcc);
+        i = i.wrapping_add(1);
+
+        let transfer = spi.dma_transfer(slave_send, slave_receive).unwrap();
+        // Bit-bang out the contents of master_send and read into master_receive
+        // as quickly as manageable. MSB first. Mode 0, so sampled on the rising
+        // edge and set on the falling edge.
+        master_cs.set_low().unwrap();
+        for (j, v) in master_send.iter().enumerate() {
+            let mut b = *v;
+            let mut rb = 0u8;
+            for _ in 0..8 {
+                if b & 128 != 0 {
+                    master_mosi.set_high().unwrap();
+                } else {
+                    master_mosi.set_low().unwrap();
+                }
+                master_sclk.set_low().unwrap();
+                b <<= 1;
+                rb <<= 1;
+                delay.delay_us(1u32);
+                master_sclk.set_high().unwrap();
+                if master_miso.is_high().unwrap() {
+                    rb |= 1;
+                }
```

Comment:
I think you should add some delay here, too. Right now you have a very asymmetric clock signal, with the low phase having a delay and the high not. The capture looks fine, but it is also only seems to be sampled at 4MHz. I actually don't understand how the capture doesn't seem to be missing clock pulses still.

Reply:
That was something I considered, but given that the data never changes on the miso line, I didn't think it was the issue at the moment.

When the clock is high, the code runs a lot more instructions than when it's low - those instructions take long enough that the 4MHz sample rate picks up two or three samples with the clock high. The cpu clock is something like 80 or 160MHz right? It's 160 on the c3 chip. 80MHz would be 40-60 cycles worth of instructions to show up for 2-3 of the LA samples. That's about right for what the code is doing - and that's why it looks ok on the capture.

I suppose the question is, with an SPI slave that's supposed to handle a 40MHz clock, is it also ok at the SPI peripheral - does the data on the MOSI line make it into the receive buffer?

On the other hand... I guess it's not that bad, either. Especially if it's limited to one or two chips.

I doubt that will fix the 0xff bytes on miso though. We'll see, I guess. :)

Reply:
> That was something I considered, but given that the data never changes on the miso line, I didn't think it was the issue at the moment.

But the MISO does change, it has some glitches when I guess the DMA can't fill the slave send data register fast enough (or the capture is noisy...?) :)

> When the clock is high, the code runs a lot more instructions than when it's low

Hmm, maybe, it's not obvious to me just by looking at the lines.

> does the data on the MOSI line make it into the receive buffer

The code doesn't consider setup and hold times, you just read as fast you can. There is some consideration you need to take - when routing SPI signals through the GPIO matrix, you need to allow more time. This is I think one of the reasons why esp-idf documentation suggests lower clock speeds are supported when using the GPIO matrix.

It looks like these parameters (setup/hold times, and even clock duty cycle requirements) aren't specified by the data sheets.